src/ChainSafe.Gaming.MetaMask.Unity/MetaMaskController.cs

```
@@ -0,0 +1,197 @@
+        {
+            var rpcRequest = new UnityRpcRequest<T>(new MetamaskWebglCoroutineRequestRpcClientFactory(ConnectedAddress));
+
+            yield return rpcRequest.SendRequest(data);
+
+            if (rpcRequest.Exception != null)
+            {
+                logger.LogError($"MetaMask Exception while making {data.Method} request {rpcRequest.Exception.Message}");
+
+                // Even if request fails we still need to callback so request task completion source result can be set.
+                onResponse?.Invoke(default);
+
+                throw rpcRequest.Exception;
+            }
+
+            logger.Log($"Successful {data.Method} JsonRPC response with result {rpcRequest.Result}");
+
+            onResponse?.Invoke(rpcRequest.Result);
+        }
+
+        /// <summary>
+        /// Callback when MetaMask successfully connects to an account.
+        /// </summary>
+        /// <param name="address">Connected Address.</param>
+        public void EthereumEnabled(string address)
+        {
+            logger.Log("Ethereum Enabled.");
+
+            if (!isInitialized)
+            {
+                // Subscribe to callbacks.
+                MetamaskWebglInterop.EthereumInit(gameObject.name, nameof(NewAccountSelected), nameof(ChainChanged));
+
+                // Get and log ChainId.
+                MetamaskWebglInterop.GetChainId(gameObject.name, nameof(ChainChanged), nameof(DisplayError));
+
+                isInitialized = true;
+            }
+
+            InvokeAccountConnected(address);
+        }
+
+        private void InvokeAccountConnected(string address)
+        {
+            OnAccountConnected?.Invoke(address);
+        }
+
+        /// <summary>
+        /// Callback when a new account is selected by user.
+        /// </summary>
+        /// <param name="address">New selected account address.</param>
+        public void NewAccountSelected(string address)
```

I don't think we should allow to switch accounts mid-game. At least right now there is no event nor nothing for the user to handle this.

---

Metamask allows i and we can't inhibit Metamasks function, what we can do it maybe throw an exception or log an error when a new account is selected

---

Yeah, that should be the way to go :)

---

done